### PR TITLE
Document the Api Schema Contract

### PR DESCRIPTION
**Pull request by solankisamir**
Based on issue https://github.com/Azure/azure-rest-api-specs/issues/1776

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (jhendrixMSFT) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: Example missing for this operation.

**Comment by solankisamir** [comment]
@ravbhatnagar FYI

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/examples/ApiManagementGetApiSchema.json
Comment (jhendrixMSFT) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/examples/ApiManagementGetApiSchema.json]: This reference cannot be resolved, see [log](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/283525487).

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/examples/ApiManagementGetApiSchema.json
Comment (jhendrixMSFT) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/examples/ApiManagementGetApiSchema.json]: This reference cannot be resolved, see [log](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/283525487).

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (jhendrixMSFT) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: Since this operation returns pageable results I believe you need to annotate it with the x-ms-pageable extension.  See [here](https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md) for more information.

**Review by jhendrixMSFT — CHANGES_REQUESTED** [review]

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/examples/ApiManagementGetApiSchema.json
Comment (solankisamir) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/examples/ApiManagementGetApiSchema.json]: That is just a Schema. It will differ from Api to API. I will update the example with a WSDL Api Schema, so that the tooling doesn't get confused.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/examples/ApiManagementGetApiSchema.json
Comment (solankisamir) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/examples/ApiManagementGetApiSchema.json]: That is just a Schema. It will differ from Api to API. I will update the example with a WSDL Api Schema, so that the tooling doesn't get confused.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (solankisamir) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: nice catch. #fixed

**Comment by jhendrixMSFT** [comment]
There are still some model validation failures for the examples, see the [log](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/284430714).  Note that you can run the model validation locally, you can find instructions for installing the validation tool [here](https://github.com/Azure/oav), use the validate-example command.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (ravbhatnagar) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: provisioningState property missing? DO you need it? Just wanted to check since you are returning 201. But its not long running so its not required. But still checking.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (ravbhatnagar) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: Is there a upper limit on the size of this?

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (ravbhatnagar) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: Also add a 200 OK response.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (ravbhatnagar) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: Is schemas a tracked or a untracked resource. If tracked PATCH would be needed. But I am guessing this is untracked resource. Please confirm

**Review by ravbhatnagar — COMMENTED** [review]
Looks fine. A few minor notes/questions

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (solankisamir) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: This is again a proxy resource in ARM. We dont need the provisioningState property.

---
In reply to: [143310129](https://github.com/Azure/azure-rest-api-specs/pull/1801#discussion_r143310129) [](ancestors = 143310129)

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (solankisamir) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: no upper size limit.

---
In reply to: [143310265](https://github.com/Azure/azure-rest-api-specs/pull/1801#discussion_r143310265) [](ancestors = 143310265)

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (solankisamir) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: We only return a 204 on Delete. This is an untracked resource in ARM.

---
In reply to: [143310351](https://github.com/Azure/azure-rest-api-specs/pull/1801#discussion_r143310351) [](ancestors = 143310351)

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (solankisamir) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: untracked resource

---
In reply to: [143310482](https://github.com/Azure/azure-rest-api-specs/pull/1801#discussion_r143310482) [](ancestors = 143310482)

**Comment by jhendrixMSFT** [comment]
Looks good from SDK perspective.  Once we have ARM sign-off we can merge.

**Comment by solankisamir** [comment]
@ravbhatnagar let me know if you have any questions. We are waiting on your sign-off.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (ravbhatnagar) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: @solankisamir - being untracked or tracked has no bearing on this. 200 on a DELETE should be the response code returned to the customers if the resource was successfully deleted. 204OK is fine to return when the requested resource was not found but the URI was well formed. This is as per the RPC and applies to both tracked and untracked resource.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (ravbhatnagar) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: Again, I think we should have a call and clarify a few things around tracked and non-tracked :). If it is long running operation, and you are returning 201 for the LRO, you will need to add a provisioning state property. Is it LRO?

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (ravbhatnagar) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: sounds good.

**Comment by ravbhatnagar** [comment]
@solankisamir - left some comments
@jhendrixMSFT - based on the answers to the 2 open questions, feel free to merge. for #1, the RP should return a 200 OK for delete. For #2, if its long running and RP is returning a 201 response, a provisioningState needs to be added. If not, then its ok.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (solankisamir) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: As per https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9.7 if the response does not return entity then 204 is expected.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (solankisamir) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: This is not a LRO.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json
Comment (ravbhatnagar) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: @solankisamir  - Yes, I am aware of that and it is one place where ARM recommendation slightly different. But as per RPC, the recommendation is as follows - 
"The resource provider can return 200 (OK) or 204 (NoContent) to indicate that the operation completed successfully. A 200 (OK) should be returned if the object exists and was deleted successfully; and a 204 (NoContent) should be used if the resource does not exist and the request is well formed."
https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#delete-resource
It will work and everything as you have it now, but for consistency across the azure platform, I think this should be fixed.

Comment (solankisamir) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: @ravbhatnagar we don't return 200 for any of our DELETE currently on this and it's parent entity and we have this behavior documented. We believe we can do this consistently across all entities with a new api-version, as it could be breaking change for some customers who have already taken dependency. What do you suggest?

Comment (jhendrixMSFT) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/2017-03-01/apimapis.json]: @solankisamir Presumably this has been the behavior for some time, so with that in mind IMO fixing this in a new API version is the correct way to go.  @ravbhatnagar are you ok with that?

**Review by jhendrixMSFT — APPROVED** [review]

**Pull request merged**
